WIP - start of README overhaul

lholmquist opened the pull request:
connects to #126

lholmquist commented:
rebased with latest master

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+Each resource gets its own file, which contains some skeleton of a resource description. Nodeshift will enrich it and then combine all the resources into a single openshift.yml and openshift.json(located in ./tmp/nodeshift/resource/).
 
-`oc login -u developer` for example.
+The resource objects `Kind`, if not given, will be extracted from the filename.
```
possessive: "object's"

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
-This guide also assumes you have a modern version(4+) of Node.js and npm installed.
 
-#### Example
+### .Nodeshift Directory
```
It's not totally clear from this description that it's the user's responsibility to supply the resource files in the `.nodeshift` directory. What is the minimum set of resource files required for a project?

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+    nodeshift -d SSO_AUTH_SERVER_URL=https://sercure-url
 
-    $ nodeshift
+__note that we left off the "${}" on the key,  nodeshift knows to search for a key with ${} added back on__
```
Not sure this is entirely necessary.

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+## Advanced Options
 
-**undeploy** - removes things that were deployed with the resource command
+While nodeshift tries to be very opinionated during the deployment process, but there are options available to pass to the cli or the API.
```
This sentence is awkward. How about

While nodeshift is very opinionated about deployment parameters, both the CLI and the API accept options that allow you to customize nodeshift's behavior.

lholmquist replied:
english is hard :)

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
 
-#### Template Parameters
+#### projectLocation
+Changes the default location of where to look for your project. Defaults to your current working directory(CWD)
```
s/of where to look//

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
 
-Some templates might need to have a value set at "run time".  For example, in the template below, we have the `${SSO_AUTH_SERVER_URL}` parameter:
+#### configLocation
+Changes the default location of where to look for a configfile. Defaults to ~/.kube/
```
s/of where to look for a configfile/of the config file/

lance replied:
What goes in the config file?

lholmquist replied:
tbh,  i don't know if we should really even let the user do this.  All that config stuff is created and managed by the oc commands(and kubectl also),

lance replied:
I agree. I don't think it is something that needs to be exposed.

lholmquist replied:
ok,  removing this

jamesfalkner replied:
I went to upgrade to latest nodeshift in [a project that runs in a Katacoda environment](https://learn.openshift.com/middleware/rhoar-getting-started-nodejs/), where the configuration file for the `oc` command is _not_ at `~/.kube/config`. It's actually at `/openshift.local.config/master/admin.kubeconfig`, and I was using the `configLocation` in nodeshift to point at it. I'm not an expert at kubernetes config locations and don't know if this is a bad practice, but removing this prevents my project from using latest nodeshift...

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
-                  - name: REALM
-                    value: master
+#### osc.strictSSL
+Setting to pass to the [Openshift Rest Client](https://www.npmjs.com/package/openshift-rest-client) for SSL use.  To allow using a self-signed cert, set to false
```
Maybe here and in `osl.tryServiceAccount` the wording could be a little clearer. It took me a few reads to understand what you mean. How about "This option is passed through to the ...".

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+Setting to pass to the [Openshift Rest Client](https://www.npmjs.com/package/openshift-rest-client) for SSL use.  To allow using a self-signed cert, set to false
 
-To set that using nodeshift, use the `-d` option with a KEY=VALUE, like this:
+#### osl.tryServiceAccount
```
Why is this prefix `osl` and the other is `osc`. Should we be consistent?

lholmquist replied:
one is for openshift-rest-client(osc)  and the other is for openshift-config-loader(osl)

really it should probably be orc and ocl

lance replied:
I guess I just wonder if it really matters for the end user. They don't care whether it's the rest client or the config loader, and they don't need to know. So, having a prefix that differs between these options doesn't seem to buy anything. Could they not both just be `os` for OpenShift?

lholmquist replied:
i think that makes sense.  I know the boosters are using that `osc.strictSSL` flag, so we need to remember to update that :)

lholmquist replied:
i'm thinking, does it even need to be `os.something` ,  if we are hiding the rest-client and configloader from the user,  then why not just make the flags `--strictSSL` and `--tryServiceAccount`

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+Specify the version of Node.js to use for the deployed application. defaults to latest.  These version tags corespond to the docker hub tags of the [bucharest-gold s2i images](https://hub.docker.com/r/bucharestgold/centos7-s2i-nodejs/tags/)
 
-__note that we left off the "${}" on the key,  nodeshift knows to search for a key with ${} added back on__
+#### quiet
```
Actually, this is an environment variable. I'm not sure why I did it this way (and maybe it should change to be a CLI option). In any case, to suppress `INFO` and `TRACE` lines, you have to `NODESHIFT_QUIET=true nodeshift`.

lholmquist replied:
I thought it was an option,  https://github.com/bucharest-gold/nodeshift/blob/master/bin/nodeshift#L63

lance replied:
You're right. I will fix it. https://github.com/bucharest-gold/nodeshift/issues/154

lance commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
 
-For more on writing openshift templates, [see here](https://docs.openshift.org/latest/dev_guide/templates.html#writing-templates)
+#### build.recreate
+Flag to recreate a buildConfig or Imagestream.  Defaults to false. Choices are "buildConfig", "imageStream", false, true
```
If `true` are both the `BuildConfig` and `ImageStream` recreated? Also, capitalize `BuildConfig` in the first sentence.

lholmquist replied:
correct, a value of true will do both,  i'll make that more clear

lholmquist commented on README.md:
```diff
@@ -1,12 +1,15 @@
 
-### Install
+* Node.js - version 8.x or greater
+* npm - version 4.x or greater
```
probably not

lholmquist commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+### Commands & Goals
 
-For this example, we will run against an Openshift Cluster running locally using Minishift.
+By default, if you run just `nodeshift`, it will run the `deploy` goal, which runs everything.
```
+1

lholmquist commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+**deploy** - does everything, the archiving of code and everything that apply-resource does
 
-These are also part of the minishift install, and [step 2 of this section](https://docs.openshift.org/latest/minishift/getting-started/quickstart.html#starting-minishift) tells you how to put them on your path.
+**undeploy** - removes things that were deployed with the apply-resource command
```
s/things/resources

lholmquist commented on README.md:
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
+OpenShift resource fragments are user provided YAML files that can be enriched with metadata, labels and more by nodeshift.
 
-Before running the nodeshift command, we first need to login to our Openshift Cluster running on Minishift using the `oc login` command
+Each resource gets its own file, which contains some skeleton of a resource description. Nodeshift will enrich it and then combine all the resources into a single openshift.yml and openshift.json(located in ./tmp/nodeshift/resource/).
```
I think the Enricher heading sort of says this

lholmquist commented:
Thanks for the review,  i hate writing readme's  :)   anyway.  i think i got all the things.

I've removed the configLocation flag and changed the osl.tryServiceAccount and osc.strictSSL flags to just be `tryServiceAccount` and `strictSSL`

lholmquist commented:
ok, now i think i fixed everything wrt the osl and osc stuff.  forgot about the public api part

lholmquist commented:
could you create a new issue so we can discuss there,  I can't seem to find
this comment on the closed PR

On Mon, Mar 12, 2018 at 1:14 PM, James Falkner <notifications@github.com>
wrote:

> *@jamesfalkner* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/bucharest-gold/nodeshift/pull/146#discussion_r173876210>
> :
>
> >
> -Some templates might need to have a value set at "run time".  For example, in the template below, we have the `${SSO_AUTH_SERVER_URL}` parameter:
> +#### configLocation
> +Changes the default location of where to look for a configfile. Defaults to ~/.kube/
>
> I went to upgrade to latest nodeshift in a project that runs in a
> Katacoda environment
> <https://learn.openshift.com/middleware/rhoar-getting-started-nodejs/>,
> where the configuration file for the oc command is *not* at ~/.kube/config.
> It's actually at /openshift.local.config/master/admin.kubeconfig, and I
> was using the configLocation in nodeshift to point at it. I'm not an
> expert at kubernetes config locations and don't know if this is a bad
> practice, but removing this prevents my project from using latest
> nodeshift...
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/bucharest-gold/nodeshift/pull/146#discussion_r173876210>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_TUtvyd1-CFWEPTQXTIZv6IZZitg2eks5tdq0DgaJpZM4Q1TRJ>
> .
>
